File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

const int LEFT = 0;
const int RIGHT = 1;

class Solution {
public:
    int res = 0;
    
    void maxZigZag(TreeNode* root, int dir, int currLength) {
        if (root == nullptr) {
            return;
        }
        res = max(res, currLength);
        maxZigZag(root->left, LEFT, (dir == RIGHT) ? currLength + 1 : 1);
        maxZigZag(root->right, RIGHT, (dir == LEFT) ? currLength + 1 : 1);
    }
    
    int longestZigZag(TreeNode* root) {
        maxZigZag(root, LEFT, 0);
        maxZigZag(root, RIGHT, 0);
        return res;
    }
};

File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public static int LEFT = -1;
    public static int RIGHT = 1;
    public int result = 0;
    
    public int longestZigZag(TreeNode root) {
        helper(root.left, 0, LEFT);
        helper(root.right, 0, RIGHT);
        return result;
    }
    
    private void helper(TreeNode root, int currentLength, int direction) {
        if (root == null) {
            result = Math.max(result, currentLength);
            return;
        }
        if (direction == LEFT) {
            helper(root.right, currentLength + 1, RIGHT);
            helper(root.left, 0, LEFT);
        } else {
            helper(root.left, currentLength + 1, LEFT);
            helper(root.right, 0, RIGHT);
        }
    }
}